### DISCUSSION TITLE
Get longitude and latitude of grid points in cube

### DISCUSSION THREAD

**Discussion by krikru** [SciTools/iris#5032] [General]
I have a netCDF file, A, in the Lambert conformal conic projection, so naturally it doesn't have any dimensions with the coordinates longitude and latitude. However, I need to extract the ranges for the longitude and latitude coordinates of the grid points, because I want to crop another netCDF file, B, which is stored on a regular grid, to a rectangular region (still on a regular grid) that is as small as possible while still covering the grid points in A completely.

Is there any way I can get the longitudes and latitudes for all points in A after loading it into an Iris cube? Or get the minimum and maximum values for the longitude and latitude coordinates of the points?

**Comment by krikru** [comment]
Considering that Iris can convert between different coordinate systems when regridding a cube, and that I can get the latitude and longitude when hovering the mouse cursor over a point on the map when plotting a slice of the cube using `iris.quickplot.pcolormesh`, surely, doing what I request here should be no problem, right?

**Comment by trexfeathers** [comment]
Hi @kriku, this is doable but actually quite complex. You'll need to use `cartopy.crs.CRS.transform_points()`, which means first getting the Cartopy CRS from the `Cube`'s coordinate system. Have you looked into this route yet?

**Reply by kriku** [reply]
@krikru cc

**Reply by krikru** [reply]
No I haven't; I didn't know you could do it in that way and I don't know how to do what you just told me. For some cubes, I seem to be able to extract the cube's coordinate system using `cube.coord_system()`, while for other cubes, I get back `None` (for example for one cube which I load from an ERA5 netCDF file, which use a regular coordinate system). Why does the `coord_system` method only give me the coordinate systems for some cubes and not for all? How can I get the coordinate systems for all cubes? And how do I get the Cartopy CRS from a cube's coordinate system once I have that?

**Reply by krikru** [reply]
@trexfeathers

**Reply by trexfeathers** [reply]
@krikru I've seen your reply, and I'm looking into it. Please accept my apologies if we're not as responsive as you need - lots of different responsibilities to balance!

**Reply by krikru** [reply]
Hi @trexfeathers, no apologies needed; no one has any responsibility to answer anyone. I just thought maybe I needed to tag you for you to be notified, but good that you had already seen my message. Sorry if I stressed you. There are a couple of other ways I think I can solve it in too (just maybe not the most elegant).

Thank you for your previous answer btw!

**Reply by trexfeathers** [reply]
Every Iris coordinate system has a `as_cartopy_crs()` method, which (as you might guess) returns the Cartopy CRS.

#### `Cube`s without a coordinate system

Iris deliberately doesn't assume a coordinate system if one isn't provided in the file. But I've looked around and if Iris is asked to plot without a coordinate system, [it uses `PlateCarree()`](https://github.com/SciTools/iris/blob/a6795cf9791597902ebb040bfe326c9aa762cbf6/lib/iris/plot.py#L944), so you could use that (`from cartopy.crs import PlateCarree`).

#### `transform_points()` syntax

Here's the definition for `cartopy.crs.CRS.transform_points()`, since there isn't online documentation for it:

```python
    def transform_points(self, src_crs, x, y, z=None, trap=False):
        """
        transform_points(src_crs, x, y[, z])

        Transform the given coordinates, in the given source
        coordinate system (``src_crs``), to this coordinate system.

        Parameters
        ----------
        src_crs
            instance of :class:`CRS` that represents the
            coordinate system of ``x``, ``y`` and ``z``.
        x
            the x coordinates (array), in ``src_crs`` coordinates,
            to transform.  May be 1 or 2 dimensional.
        y
            the y coordinates (array), in ``src_crs`` coordinates,
            to transform.  Its shape must match that of x.
        z: optional
            the z coordinates (array), in ``src_crs`` coordinates, to
            transform.  Defaults to None.
            If supplied, its shape must match that of x.
        trap
            Whether proj errors for "latitude or longitude exceeded limits" and
            "tolerance condition error" should be trapped.

        Returns
        -------
            Array of shape ``x.shape + (3, )`` in this coordinate system.

        """
```

**Reply by krikru** [reply]
@trexfeathers Thank you very much for your reply! I'm currently implementing this solution.

Do you think it would be possible to add a function to the Iris API that provided `PlateCarree()` as the default Cartopy proj? That is,

```
def get_default_cartopy_proj():
    return ccrs.PlateCarree()
```

and the change [line 944 in plot.py](https://github.com/SciTools/iris/blob/a6795cf9791597902ebb040bfe326c9aa762cbf6/lib/iris/plot.py#L944) to

```
        cartopy_proj = get_default_cartopy_proj()
```

? Or maybe even a convenience function for getting the cartopy projection from an object (be it a cube or a coordinate) that would return `PlateCarree()` if the object had no coordinate system? Then another function could be added to get the Cartopy CRS from an object.

This would effectively move the responsibility of knowing which projection to use when no coordinate system is provided from the user to Iris.

**Reply by krikru** [reply]
@trexfeathers Hm, I seem to be running into problems. I get

```
AttributeError: 'PlateCarree' object has no attribute 'as_cartopy_crs'
```

So, do you have any idea how to get the CRS from `PlateCarree()` or the CRS from a cube that doesn't have any coordinate system?

**Reply by krikru** [reply]
I see that the [Projection](https://scitools.org.uk/cartopy/docs/latest/reference/generated/cartopy.crs.Projection.html) class has a method called `source_crs`; I will try that.

**Reply by trexfeathers** [reply]
> @trexfeathers Hm, I seem to be running into problems. I get
> 
> ```
> AttributeError: 'PlateCarree' object has no attribute 'as_cartopy_crs'
> ```
> 
> So, do you have any idea how to get the CRS from `PlateCarree()` or the CRS from a cube that doesn't have any coordinate system?

The `PlateCarree` you have imported already is a Cartopy CRS, so you do not need any 'conversion'.

**Reply by krikru** [reply]
Thanks! I think I have succeeded now to extract the latitudes and longitudes for all grid points in the cube that is on a Lambert conformal conic projection.

However, I see that `PlateCaree()` have the coordinates `Easting`, `Northing` and `Ellipsoidal height` (so that's what I get out of `transform_points`), while the cube that is on a regular grid, for which `coord_system()` returns `None`, has the coordinates `latitude` and `longitude`, which firstly don't match when considering the number of coordinates, and secondly have opposite orders of the horizontal coordinates (`latitude` and `longitude` corresponds to `Northing` and `Easting`, respectively, and not to `Easting` and `Northing`), but it still plots correctly when calling `iris.quickplot.pcolormesh`. Do you know how that works? I need to map `latitude` and `longitude` to `Northing` and `Easting`, respectively, but am a bit reluctant to hardcoding this mapping into my code. It seems like Iris already must do this mapping somehow.

**Reply by trexfeathers** [reply]
>Do you know how that works?

Unfortunately my knowledge is limited when it comes to different coordinate systems and projections, sorry.

You could have a look at [the full range of Iris coordinate systems](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris/coord_systems.html) and see whether any of them seem appropriate to you.

Alternatively look directly at the range of [Cartopy CRS's](https://scitools.org.uk/cartopy/docs/latest/reference/crs.html). Full detail on this is available in [the PROJ documentation](https://proj.org/operations/projections/index.html).

Hopefully these resources help you find what you need.

**Comment by trexfeathers** [comment]
https://github.com/SciTools/iris/discussions/5032#discussioncomment-4056180

I am not keen for Iris to be making more decisions about the 'correct' coordinate system to use when one is missing; it is important that the user knows what is being done, and therefore it is important for the user to make the decision about the coordinate system. That will be why it is not auto populated when a `Cube` is loaded. Indeed I'm a little uncomfortable that Iris makes that default decision when plotting, but I don't know the history behind this choice.

Of course that's my opinion. @SciTools/iris-devs might think differently.
